### PR TITLE
Timebox multiple requests

### PR DESCRIPTION
**Pull request by doctorpod**

### CODE REVIEW

File: spec/services/timebox_spec.rb [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+  let(:result) do
+    subject.run(-> { :fellback }) do
+      sleep 0.2
+      :finished
+    end
+  end
+
+  context '#run' do
+    context 'big timebox' do
+      let(:seconds) { 10 }
```
Comment (alan) [inline spec/services/timebox_spec.rb]: Are there other ways we could test this without sleeping 10 seconds?

Comment (doctorpod) [inline spec/services/timebox_spec.rb]: We don't actually wait 10 seconds - just the 0.2 seconds in the main block (line 8) - the 10 seconds is the param we put in when we instantiate Timebox to ensure we'll never trigger a fallback.

Comment (alan) [inline spec/services/timebox_spec.rb]: 👍

File: app/services/timebox.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
+# An object that runs a provided block within the time limit or else returns
+# the result of the fallback block. run may be called multiple times; the
+# original time limit being applied to all runs collectively.
+class Timebox
+  def initialize(time_limit)
+    @deadline = Time.now.to_i + time_limit
+  end
+
+  def run(fallback_block)
+    time_remaining = @deadline - Time.now.to_i
```
Comment (mloughran) [inline app/services/timebox.rb]: Using `to_i` on the time will lead to very inconsistent timeouts being applied in practice, depending on the fractional component of the time when the deadline and current time are recorded.

File: app/controllers/api/slots_controller.rb [hunk 3]
```diff
@@ -1,10 +1,16 @@
...
     def index
       prison = Prison.enabled.find(params.require(:prison_id))
-      @slots = SlotAvailability.new(
-        prison, prisoner_number, date_of_birth, start_date..end_date
-      ).slots
+      slot_availability = SlotAvailability.new(
+        prison, prisoner_number, date_of_birth, start_date..end_date)
+      timebox = Timebox.new(TIMEBOX_LIMIT)
+
+      @slots = timebox.run(-> { slot_availability.all_slots }) {
```
Comment (alan) [inline app/controllers/api/slots_controller.rb]: Would be good to add a controller test to check the fallback response

**Pull request merged**